File: src/main/scala/com/github/unchama/seichiassist/minestack/MineStackUsageHistory.java
package com.github.unchama.seichiassist.minestack;

import java.util.ArrayList;
import java.util.List;

/**
 * Created by karayuu on 2018/06/13
 */
public final class MineStackUsageHistory {
    private static final int LIST_MAX_SIZE = 27;

    public List<MineStackObj> usageHistory = new ArrayList<>();

    /**
     * 履歴に追加します。ただし、データの保存可能な最大値を超えていた場合、先頭から削除されます。
     */
    public void add(MineStackObj obj) {
        usageHistory.remove(obj);
        usageHistory.add(obj);
        if (usageHistory.size() > LIST_MAX_SIZE) {
            usageHistory.remove(0);
        }
    }
}

File: src/main/scala/com/github/unchama/seichiassist/minestack/MineStackUsageHistory.scala
package com.github.unchama.seichiassist.minestack

class MineStackUsageHistory {

  private val maxListSize = 27

  private var _usageHistory: Vector[MineStackObject] = Vector.empty

  /**
   * アイテムを履歴に追加します。
   * ただし、追加したことで履歴の最大件数を超えた場合、時系列順で見て最も最初に追加された履歴が削除されます。
   */
  def addHistory(mineStackObject: MineStackObject): Unit = {
    _usageHistory = _usageHistory.filterNot(_ == mineStackObject)
    _usageHistory :+= mineStackObject
    if (_usageHistory.size > maxListSize) {
      _usageHistory = _usageHistory.drop(1)
    }
  }

  def usageHistory: Vector[MineStackObject] = _usageHistory

}
